Python/Product/PythonTools/PythonTools/Project/PythonProjectNode.cs

```diff
@@ -229,36 +237,32 @@ internal partial class PythonProjectNode :
                 Debug.Assert(this.FileName != null);
                 var oldActive = _active;
 
+                // stop listening for installed files changed
                 var oldPms = _activePackageManagers;
                 _activePackageManagers = null;
-
                 foreach (var pm in oldPms.MaybeEnumerate()) {
                     pm.InstalledFilesChanged -= PackageManager_InstalledFilesChanged;
                 }
 
                 lock (_validFactories) {
-                    if (_validFactories.Count == 0) {
-                        // No factories, so we must use the global default.
+                    // if there are no valid factories,
+                    // or the specified factory isn't in the valid list,
+                    // use the global default
+                    if (_validFactories.Count == 0 || value == null || !_validFactories.Contains(value.Configuration.Id)) {
                         _active = null;
-                    } else if (value == null || !_validFactories.Contains(value.Configuration.Id)) {
-                        // Choose a factory and make it our default.
```

Comment:
The change you made makes more sense to me, I don't get why previously we just randomly return the last item in the ordered env list and make it active.